### PR TITLE
[SPARK-40281][PYTHON] Memory Profiler on Executors

### PR DESCRIPTION
**Pull request by xinrong-meng**
### What changes were proposed in this pull request?
Implement PySpark memory profiling on executors. The feature is enabled via a newly-introduced Spark configuration `spark.python.profile.memory`.

See more [design.](https://docs.google.com/document/d/e/2PACX-1vR2K4TdrM1eAjNDC1bsflCNRH67UWLoC-lCv6TSUVXD91Ruksm99pYTnCeIm7Ui3RgrrRNcQU_D8-oh/pub)

### Why are the changes needed?
There are many factors in a PySpark program’s performance. Memory, as one of the key factors of a program’s performance, had been missing in PySpark profiling. A PySpark program on the Spark driver can be profiled with [Memory Profiler](https://www.google.com/url?q=https://pypi.org/project/memory-profiler/&sa=D&source=editors&ust=1668027860192689&usg=AOvVaw1t4LRcObEGuhaTr5oHEUwU) as a normal Python process, but there was not an easy way to profile memory on Spark executors.

PySpark UDFs, one of the most popular Python APIs, enable users to run custom code on top of the Apache Spark™ engine. However, it is difficult to optimize UDFs without understanding memory consumption.

The PR proposes to introduce the PySpark memory profiler, which profiles memory on executors. It provides information about total memory usage and pinpoints which lines of code in a UDF attribute to the most memory usage. That will help optimize PySpark UDFs and reduce the likelihood of out-of-memory errors.

### Does this PR introduce _any_ user-facing change?
No changes to existing user-facing behaviors.

A Spark configuration `spark.python.profile.memory` is introduced to enable the PySpark memory profiling feature.

### How was this patch tested?
- Unit tests.
- Manual tests on Jupyter notebooks as shown below:
image

### CODE REVIEW

**Comment by mridulm** [comment]
+CC @zhouyejoe

**Review by HyukjinKwon — APPROVED** [review]
i already reviewed this internally so LGTM but I will leave it some more days for others to reivew.

**Review by HyukjinKwon — COMMENTED** [review]
We should probably have some docs at https://spark.apache.org/docs/latest/api/python/development/debugging.html BTW. We can do separately.

**Comment by xinrong-meng** [comment]
Thank you @HyukjinKwon ! I will file a separate PR for the comprehensive documents.

**Comment by xinrong-meng** [comment]
@ueshin May I ask for your review? Thanks!

**Comment by Tagar** [comment]
cc @LucaCanali

File: python/pyspark/tests/test_memory_profiler.py [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+try:
+    import memory_profiler  # type: ignore[import] # noqa: F401
+
+    has_memory_profiler = True
+except Exception:
+    has_memory_profiler = False
+
+
+@unittest.skipIf(not has_memory_profiler, "Must have memory-profiler installed.")
+class MemoryProfilerTests(PySparkTestCase):
```
[the remark this answers is not in the document]
Comment (xinrong-meng) [reply python/pyspark/tests/test_memory_profiler.py:43]: Good catch! Updated.

**Review by ueshin — APPROVED** [review]
LGTM.

**Comment by HyukjinKwon** [comment]
Merged to master.

**Pull request closed without merging**